hummingbot/connector/exchange/ndax/ndax_order_book_message.py

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env python
+
+from collections import namedtuple
+from typing import (
+    Dict,
+    List,
+    Optional,
+)
+
+from hummingbot.core.data_type.order_book_row import OrderBookRow
+from hummingbot.core.data_type.order_book_message import (
+    OrderBookMessage,
+    OrderBookMessageType,
+)
+
+NdaxOrderBookEntry = namedtuple("NdaxOrderBookEntry", "mdUpdateId accountId actionDateTime actionType lastTradePrice orderId price productPairCode quantity side")
+NdaxTradeEntry = namedtuple("NdaxTradeEntry", "tradeId productPairCode quantity price order1 order2 tradeTime direction takerSide blockTrade orderClientId")
+
+
+class NdaxOrderBookMessage(OrderBookMessage):
+    def __new__(
+        cls,
+        message_type: OrderBookMessageType,
+        content: Dict[str, any],
+        timestamp: Optional[float] = None,
+        *args,
+        **kwargs,
+    ):
+        if timestamp is None:
+            if message_type is OrderBookMessageType.SNAPSHOT:
+                raise ValueError("timestamp must not be None when initializing snapshot messages.")
+            timestamp = content["timestamp"]
+
+        return super(NdaxOrderBookMessage, cls).__new__(
+            cls, message_type, content, timestamp=timestamp, *args, **kwargs
+        )
+
+    @property
+    def update_id(self) -> int:
+        if self.type in [OrderBookMessageType.DIFF, OrderBookMessageType.SNAPSHOT]:
+            entry: NdaxOrderBookEntry = self.content["data"][0]
+            return int(entry.actionDateTime)
+        elif self.type == OrderBookMessageType.TRADE:
+            entry: NdaxTradeEntry = self.content["data"][0]
+            return int(entry.tradeTime)
+
+    @property
+    def trade_id(self) -> int:
+        entry: NdaxTradeEntry = self.content["data"][0]
+        return entry.tradeId
+
+    @property
+    def trading_pair(self) -> str:
+        return self.content["trading_pair"]
+
+    @property
+    def asks(self) -> List[OrderBookRow]:
+        entries: List[NdaxOrderBookEntry] = self.content["data"]
+
+        return [
+            OrderBookRow(float(entry.price), float(entry.quantity), self.update_id)
+            for entry in entries if entry.side == 1
+        ]
+
+    @property
+    def bids(self) -> List[OrderBookRow]:
+        entries: List[NdaxOrderBookEntry] = self.content["data"]
+
+        return [
+            OrderBookRow(float(entry.price), float(entry.quantity), self.update_id)
+            for entry in entries if entry.side == 0
+        ]
+
+    def __eq__(self, other) -> bool:
```

Comment:
It might be a good idea to implement __eq__ by validating first that the classes of both objects is the same. If not the code might try to send messages to an object that does not know how to reply to them.

Also, if we reimplement __eq__ we should reimplement also __hash__ to match. Any object that is equal to other should return the same hash.

Reply:
I agree with the first that we should include a validation on the classes of both objects. As for the `__hash__` I think I will use the `OrderMessageType`, `timestamp` as well as `content` of the message to calculate the hash